File: android/src/main/java/com/reactnativeecommpay/utils/RecurrentInfoUtility.java
package com.reactnativeecommpay.utils;

import com.ecommpay.sdk.ECMPRecurrentInfo;
import com.ecommpay.sdk.ECMPRecurrentInfoSchedule;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;


public class RecurrentInfoUtility {
  public static ECMPRecurrentInfo bind(ReadableMap recurrent, ReadableArray schedules) {
    String type = recurrent.hasKey("type") ? recurrent.getString("type") : "R";
    String expiryDay = recurrent.hasKey("expiryDay") ? recurrent.getString("expiryDay") : null;
    String expiryMonth = recurrent.hasKey("expiryMonth") ? recurrent.getString("expiryMonth") : null;
    String expiryYear = recurrent.hasKey("expiryYear") ? recurrent.getString("expiryYear") : null;
    String period = recurrent.hasKey("period") ? recurrent.getString("period") : null;
    String time = recurrent.hasKey("time") ? recurrent.getString("time") : null;
    String startDate = recurrent.hasKey("startDate") ? recurrent.getString("startDate") : null;
    String scheduledPaymentID = recurrent.hasKey("scheduledPaymentID") ? recurrent.getString("scheduledPaymentID") : null;

    ECMPRecurrentInfo recurrentInfo;

    if (TypeUtility.stringIsNotEmpty(expiryDay) && TypeUtility.stringIsNotEmpty(expiryMonth) &&
      TypeUtility.stringIsNotEmpty(expiryYear) && TypeUtility.stringIsNotEmpty(period) &&
      TypeUtility.stringIsNotEmpty(time) && TypeUtility.stringIsNotEmpty(startDate) &&
      TypeUtility.stringIsNotEmpty(scheduledPaymentID)
    ) {
      recurrentInfo = new ECMPRecurrentInfo(
        type,
        expiryDay,
        expiryMonth,
        expiryYear,
        period,
        time,
        startDate,
        scheduledPaymentID
      );
    } else {
      recurrentInfo = new ECMPRecurrentInfo(
        type
      );
    }

    if (recurrent.hasKey("amount")) {
      recurrentInfo.setAmount(TypeUtility.convertToLong(recurrent.getDouble("amount")));
    }

    if (schedules.size() > 0) {
      ECMPRecurrentInfoSchedule[] ecmpRecurrentInfoSchedules = new ECMPRecurrentInfoSchedule[schedules.size()];
      for (int i = 0; i < schedules.size(); i++) {
        ReadableMap map = schedules.getMap(i);

        if (map == null) {
          continue;
        }

        String date = map.hasKey("date") ? map.getString("date") : null;
        long amount = map.hasKey("amount") ? TypeUtility.convertToLong(map.getDouble("amount")) : -1;

        if (TypeUtility.stringIsEmpty(date) || amount == -1) {
          continue;
        }

        ecmpRecurrentInfoSchedules[i] = new ECMPRecurrentInfoSchedule(date, amount);
      }

      if (ecmpRecurrentInfoSchedules.length > 0) {
        recurrentInfo.setSchedule(ecmpRecurrentInfoSchedules);
      }
    }

    return recurrentInfo;
  }
}

File: ios/Utils/RecurrentInfoUtility.swift
//
//  PaymentInfo.swift
//  Ecommpay
//
//  Created by Станислав Полещук on 22.07.2022.
//  Copyright © 2022 Facebook. All rights reserved.
//
import Foundation
import ecommpaySDK

public class RecurrentInfoUtility {
  static func bind(recurrent: NSDictionary, schedules: NSArray) -> RecurrentInfo {

    var type = RecurrentType.OneClick

    let typeStr = recurrent["type"] as? String

    switch typeStr {
    case "R":
      do {
        type = RecurrentType.Regular
        break
      }
    case "C":
      do {
        type = RecurrentType.OneClick
        break
      }
    case "U":
      do {
        type = RecurrentType.Autopayment
        break
      }
    default:
      do {
        type = RecurrentType.Autopayment
      }
    }

    let expiryDay = recurrent["expiryDay"] as? String ?? nil
    let expiryMonth = recurrent["expiryMonth"] as? String ?? nil
    let expiryYear = recurrent["expiryYear"] as? String ?? nil
    let period = recurrent["period"] as? ecommpaySDK.RecurrentPeriod ?? nil
    let time = recurrent["time"] as? String ?? nil
    let startDate = recurrent["startDate"] as? String ?? nil
    let scheduledPaymentID = recurrent["scheduledPaymentID"] as? String ?? nil
    let amount = recurrent["amount"] as? Int ?? nil

    let recurrentInfo: RecurrentInfo

    if expiryDay != nil && expiryMonth != nil && expiryYear != nil && period != nil && time != nil
      && startDate != nil && scheduledPaymentID != nil
    {
      recurrentInfo = RecurrentInfo(
        type: type,
        expiryDay: expiryDay!,
        expiryMonth: expiryMonth!,
        expiryYear: expiryYear!,
        period: period!,
        time: time!,
        startDate: startDate!,
        scheduledPaymentID: scheduledPaymentID!
      )
    } else {
      recurrentInfo = RecurrentInfo(
        type: type
      )
    }

    if amount != nil {
      recurrentInfo.setAmount(amount: amount!)
    }

    if schedules.count > 0 {
      var arrSchedule = [ecommpaySDK.RecurrentInfoSchedule]()

      schedules.forEach { (x) in
        let shedule = x as AnyObject

        print(shedule)
        let date = shedule["date"]
        let amount = shedule["amount"]

        if date != nil && amount != nil {
          arrSchedule.append(RecurrentInfoSchedule(date: date as! String, amount: amount as! Int))
        }
      }

      recurrentInfo.setSchedule(schedule: arrSchedule)
    }

    return recurrentInfo
  }
}
